File: concurrency/P/Print Zero Even Odd/Print Zero Even Odd.cpp
class ZeroEvenOdd {
private:
    int n;
    mutex mx;
    condition_variable cv;
    atomic<int> num_to_print;
    atomic<int> current;

public:
    ZeroEvenOdd(int n) {
        this->n = n;
        num_to_print = 0;
        current = 0;
    }

    // printNumber(x) outputs "x", where x is an integer.
    void zero(function<void(int)> printNumber) {
        do_work(printNumber, [&]{return num_to_print == 0;});
    }

    void even(function<void(int)> printNumber) {
        do_work(printNumber, [&]{return num_to_print != 0 && num_to_print % 2 == 0;});
    }

    void odd(function<void(int)> printNumber) {
         do_work(printNumber, [&]{return num_to_print % 2 != 0;});
    }
    
protected:
   void do_work(function<void(int)> print, function<bool()> eval)
    {
        while(current <= n)
        {
            unique_lock<mutex> ul(mx);
            cv.wait(ul, [&]{return eval() || current > n;});
            if(current > n) break;
            print(num_to_print);
            num_to_print = num_to_print == 0 || num_to_print == n ? ++current : 0;
            cv.notify_all();
        }
    }
};

File: concurrency/P/Print Zero Even Odd/Print Zero Even Odd.py
from threading import Semaphore
class ZeroEvenOdd:
    def __init__(self, n):
        self.n = n
        self.semOdd  = Semaphore(0) # Permission to write an Odd  Number
        self.semEven = Semaphore(0) # Permission to write an Even Number
        self.semZero = Semaphore(1) # Permission to write a  Zero
	# printNumber(x) outputs "x", where x is an integer.
    def zero(self, printNumber):
        for i in range(1,self.n+1):
            # A) Request Permission to Write a Zero
            self.semZero.acquire()
            printNumber(0)
            # B) Check if "i" if Odd or Even, and give permission to write the number
            if i&1:
                self.semOdd.release()
            else:
                self.semEven.release()
            # C) Permission to write a zero is returned by another Thread ( release triggers a change)
    def even(self, printNumber):
        # A) Iterate only through Even numbers
        for i in range(2,self.n+1,2):
            # B) Request Permission to Write Current Number
            self.semEven.acquire()
            printNumber(i)
            # C) Return Permission to Write a Zero (if applicable)
            self.semZero.release()
    def odd(self, printNumber):
        # A) Iterate only through Odd numbers
        for i in range(1,self.n+1,2):
            # B) Request Permission to Write Current Number
            self.semOdd.acquire()
            printNumber(i)
            # C) Return Permission to Write a Zero (if applicable)
            self.semZero.release()
